src/gf.c

```diff
@@ -2378,29 +2389,37 @@ jl_code_instance_t *jl_compile_method_internal(jl_method_instance_t *mi, size_t
                 codeinst2->min_world, codeinst2->max_world);
         if (jl_atomic_load_relaxed(&codeinst->invoke) == NULL) {
             codeinst->rettype_const = codeinst2->rettype_const;
-            uint8_t specsigflags = jl_atomic_load_acquire(&codeinst2->specsigflags);
             jl_callptr_t invoke = jl_atomic_load_acquire(&codeinst2->invoke);
             void *fptr = jl_atomic_load_relaxed(&codeinst2->specptr.fptr);
             if (fptr != NULL) {
-                while (!(specsigflags & 0b10)) {
-                    jl_cpu_pause();
-                    specsigflags = jl_atomic_load_acquire(&codeinst2->specsigflags);
-                }
+                _jl_specsig_flags_t specsig = wait_for_code_instance(codeinst2);
                 invoke = jl_atomic_load_relaxed(&codeinst2->invoke);
                 void *prev_fptr = NULL;
                 // see jitlayers.cpp for the ordering restrictions here
                 if (jl_atomic_cmpswap_acqrel(&codeinst->specptr.fptr, &prev_fptr, fptr)) {
-                    jl_atomic_store_relaxed(&codeinst->specsigflags, specsigflags & 0b1);
+                    _jl_specsig_flags_t specialized_flag_only = { {
+                        /* specptr_specialized */ specsig.flags.specptr_specialized,
+                        /* specptr_matches_invokeptr */ 0,
+                        /* from_image */ 0
+                    } };
+                    _jl_specsig_flags_t invokeptr_matches_flag_only = { {
+                        /* specptr_specialized */ 0,
+                        /* specptr_matches_invokeptr */ specsig.flags.specptr_matches_invokeptr,
+                        /* from_image */ 0
+                    } };
+
+                    jl_atomic_store_relaxed(&codeinst->specsigflags, specialized_flag_only.bits);
                     jl_atomic_store_release(&codeinst->invoke, invoke);
-                    jl_atomic_store_release(&codeinst->specsigflags, specsigflags);
-                } else {
+                    // unspec is probably not specsig, but might be using specptr
```

Comment:
This might want to set `invoke_matches_specptr` to be `unspec->specptr != nullptr`, since we know that we can't be using specptr if specptr is nullptr.

Reply:
This bit is supposed to be set after an invoke pointer that calls the specptr is written to the codeinst. It's basically just a barrier to force reads of the specptr and reads of the invoke pointer to match up correctly. Codegen shouldn't be relying on that bit beyond using it as a barrier (if it is, it should be doing non-null checks on specptr instead).

Reply:
Codegen is required to respect this bit to know if the specptr is valid for it to use, or whether only the invoke function is permitted to use this field.